library/src/test/java/org/apache/apex/malhar/lib/dedup/DeduperPartitioningTest.java

```diff
@@ -0,0 +1,171 @@
+    public void populateDAG(DAG dag, Configuration conf)
+    {
+      TestGenerator gen = dag.addOperator("Generator", new TestGenerator());
+
+      TestDeduper dedup = dag.addOperator("Deduper", new TestDeduper());
+      dedup.setKeyExpression("id");
+      dedup.setTimeExpression("eventTime.getTime()");
+      dedup.setBucketSpan(60);
+      dedup.setExpireBefore(600);
+      
+      ConsoleOutputOperator console = dag.addOperator("Console", new ConsoleOutputOperator());
+      dag.addStream("Generator to Dedup", gen.output, dedup.input);
+      dag.addStream("Dedup to Console", dedup.unique, console.input);
+      dag.setInputPortAttribute(dedup.input, Context.PortContext.TUPLE_CLASS, TestEvent.class);
+      dag.setOutputPortAttribute(dedup.unique, Context.PortContext.TUPLE_CLASS, TestEvent.class);
+      dag.setAttribute(dedup, Context.OperatorContext.PARTITIONER, 
+          new StatelessPartitioner<TimeBasedDedupOperator>(NUM_DEDUP_PARTITIONS));
+    }
+  }
+
+  public static class TestDeduper extends TimeBasedDedupOperator
+  {
+    int operatorId;
+    boolean started = false;
+    HashMap<Integer, Integer> partitionMap = Maps.newHashMap();
+
+    @Override
+    public void setup(OperatorContext context)
+    {
+      super.setup(context);
+      operatorId = context.getId();
+    }
+
+    @Override
+    protected void processTuple(Object tuple)
+    {
+      TestEvent event = (TestEvent)tuple;
+      if (partitionMap.containsKey(event.id)) {
+        if (partitionMap.get(event.id) != operatorId) {
+          testFailed = true;
+          throw new RuntimeException("Wrong tuple assignment");
+        }
+      } else {
+        partitionMap.put(event.id, operatorId);
+      }
+    }
+  }
+
+  public static class TestGenerator extends BaseOperator implements InputOperator
+  {
+
+    public final transient DefaultOutputPort<TestEvent> output = new DefaultOutputPort<>();
+    private final transient Random r = new Random();
+
+    @Override
+    public void emitTuples()
+    {
+      TestEvent event = new TestEvent();
+      event.id = r.nextInt(100);
+      output.emit(event);
+    }
+  }
+
+  public static class TestEvent
+  {
+    private int id;
+    private Date eventTime;
+
+    public TestEvent()
+    {
+    }
+
+    public int getId()
+    {
+      return id;
+    }
+
+    public void setId(int id)
+    {
+      this.id = id;
+    }
+
+    public Date getEventTime()
+    {
+      return eventTime;
+    }
+
+    public void setEventTime(Date eventTime)
+    {
+      this.eventTime = eventTime;
+    }
+  }
+
+  /**
+   * This test validates whether a tuple key goes to exactly one partition
+   */
+  @Test
+  public void testDeduperStreamCodec()
+  {
+    try {
+      LocalMode lma = LocalMode.newInstance();
+      Configuration conf = new Configuration(false);
+      lma.prepareDAG(new TestDedupApp(), conf);
+      LocalMode.Controller lc = lma.getController();
+      lc.run(10 * 1000); // runs for 10 seconds and quits
```

Comment:
Can you run the test in async mode of Controller using method similar to:
com.datatorrent.lib.io.fs.FileSplitterBaseTest.testSplitterInApp

This gives a deterministic way of verification the cases.